Task6/arrayFunctionCollection.js

```diff
@@ -0,0 +1,116 @@
+const choosingFunc = function()  {
+    //arguments[0] - parametrs (Array*param+)
+    //arguments[1] - function     
+	if( Array.isArray(arguments[0][0]))  {
+        var parameters = Array.prototype.slice.call(arguments[0], 1);
+		const resultArray = arguments[1].call(this, parameters, arguments[0][0]);
+		return resultArray;
+	}
+	const resultFunc = arguments[1].bind(this, ...arguments[0]);
+    this._func[this._func.length] = resultFunc;
+	return this;
+}
+
+var arrayFunctionCollection = {
+
+    _array : [], 
+    _func : [],
+
+	take : function() {
+		return choosingFunc.call(this,
+			arguments,
+			function ( number, array){
+				let newArr = [];
+				for(let i = 0; (i < number && i < array.length); i++){
+					newArr[i] = array[i];
+				}
+				return newArr;
+			}, 			
+		);
+    },
+    
+    skip : function (){
+		return choosingFunc.call(this,
+			arguments,
+			function ( number, array){
+				let newArr = [];
+				for(let i = number; i < array.length; i++){
+					newArr[i - number] = array[i];
+				}
+				return newArr;
+			},
+		);
+    },	
+
+    map :  function (){
+		return choosingFunc.call(this, 
+			arguments, 
+			function (callback, array){
+				for(let i = 0; i < array.length; i++){
+					array[i] = callback(array[i], i, array);
+				}
+				return array;
+			},
+		);
+	},	
+
+	reduce : function ( callback, initialValue, array){
+		let result = initialValue;
+		for(let i = 0; i < array.length; i++){
+			result = callback(result, array[i]);
+		}
+		return result;
+	},
+
+	filter : function(){ 
+		return choosingFunc.call(
+			this, 
+			arguments, 
+			function (callback, array){
+				let newArr = [];
+				let j = 0;
+				for(let item of array){
+					if (callback(item)){
+						newArr[j] = item;
+						j++;
+					}
+				}
+				return newArr;
+			}
+		);
+	},
+
+	foreach : function(){ 
+		return choosingFunc.call(
+			this, 
+			arguments, 
+			function ( callback, array){
+				for(let i = 0; i < array.length; i++){
+					array[i] = callback(array[i]);
+				}
+				return array;
+			}
+		);
+	},
+
+	chain: function (array){
```

Comment:
In the current approach I see an issue that every `chain` call overwrite array (if `value` hasn't been called). Let's discuss better solution.